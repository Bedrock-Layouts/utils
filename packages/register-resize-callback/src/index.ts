export type ResizeFunc = (r: ResizeObserverEntry) => unknown;
export type DeregisterFunc = () => unknown | Error;

type ObserverCallback =
  | ResizeFunc
  | {
      current: ResizeFunc;
    };

// eslint-disable-next-line functional/no-let
let observer: ResizeObserver | undefined;

const callBackMap = new WeakMap<Element, ObserverCallback[]>();

export function init(): ResizeObserver {
  /* c8 ignore next */
  if (observer) return observer;

  return (observer = new ResizeObserver((entries) => {
    return entries.map((entry) => {
      const maybeCallBacks = callBackMap.get(entry.target);
      /**
       * This is defensive code, but it's possible that get returns undefined
       */
      /* c8 ignore next */
      const safeCallBacks = maybeCallBacks ?? [];

      if (safeCallBacks.length === 0) return entry;

      requestAnimationFrame(() => {
        return safeCallBacks.map((callback) => {
          const safeCallback =
            typeof callback === "function" ? callback : callback.current;

          safeCallback(entry);

          return callback;
        });
      });
      return entry;
    });
  }));
}

export function registerCallback(
  node: Readonly<Element>,
  // eslint-disable-next-line functional/prefer-immutable-types
  cb: ObserverCallback,
): DeregisterFunc {
  if (observer === undefined) {
    init();
  }

  const callbacks = callBackMap.get(node) ?? [];

  callBackMap.set(node, callbacks.concat(cb));
  observer?.observe(node);

  return () => {
    const maybeCallBacks = callBackMap.get(node);
    /**
     * This is defensive code, but it's possible that get returns undefined
     */
    /* c8 ignore next */
    const safeCallBacks = maybeCallBacks ?? [];
    const newCallbacks = safeCallBacks.filter((ref) => ref !== cb);

    callBackMap.set(node, newCallbacks);

    if (newCallbacks.length === 0) {
      observer?.unobserve(node);
      callBackMap.delete(node);
    }
  };
}
