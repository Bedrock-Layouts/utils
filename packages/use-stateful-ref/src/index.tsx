import React from "react";

export function useStatefulRef<T>(initialVal?: T): React.MutableRefObject<T> {
  // eslint-disable-next-line prefer-const, functional/no-let
  let [cur, setCur] = React.useState<T | undefined>(initialVal);

  const { current: ref } = React.useRef({
    current: cur,
  });

  // eslint-disable-next-line functional/immutable-data
  Object.defineProperty(ref, "current", {
    get: () => cur as T,
    // eslint-disable-next-line functional/no-return-void
    set: (value: T) => {
      if (!Object.is(cur, value)) {
        cur = value;
        setCur(value);
      }
    },
  });

  return ref as React.MutableRefObject<T>;
}
