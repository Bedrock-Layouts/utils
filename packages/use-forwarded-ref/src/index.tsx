import { useStatefulRef } from "@bedrock-layout/use-stateful-ref";
import { useRef } from "react";
import React from "react";

export interface Config {
  isStateful: boolean;
}

export function useForwardedRef<T>(
  forwardedRef?: React.Ref<T>,
  config: Config = { isStateful: true },
): React.MutableRefObject<T> {
  const statefulRef = useStatefulRef<T>();
  const ref = useRef<T>();

  const innerRef = config.isStateful ? statefulRef : ref;

  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-expect-error
  React.useImperativeHandle(forwardedRef, () => innerRef.current);

  return innerRef as React.MutableRefObject<T>;
}
