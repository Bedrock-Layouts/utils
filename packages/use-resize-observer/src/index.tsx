import {
  ResizeFunc,
  init,
  registerCallback,
} from "@bedrock-layout/register-resize-callback";
import { useStatefulRef } from "@bedrock-layout/use-stateful-ref";
import { useEffect, useLayoutEffect, useRef } from "react";

export function useResizeObserver<T extends Element>(
  callback: ResizeFunc,
  node?: T,
): React.MutableRefObject<T> {
  const callbackRef = useRef<ResizeFunc>(callback);

  const nodeRef = useStatefulRef<T>();

  // eslint-disable-next-line functional/no-return-void
  useLayoutEffect(() => {
    // eslint-disable-next-line functional/immutable-data
    callbackRef.current = callback;
  });

  // eslint-disable-next-line functional/no-return-void
  useEffect(() => {
    init();
  }, []);

  const currentNode = nodeRef.current ?? node;

  useEffect(() => {
    init();

    /**
     * node is undefined when the component is unmounted or not yet mounted
     */
    /* c8 ignore start */
    const cleanup = currentNode
      ? registerCallback(currentNode, callbackRef)
      : () => 0;
    /* c8 ignore end */

    // eslint-disable-next-line functional/no-return-void
    return () => {
      cleanup();
    };
  }, [currentNode]);

  return nodeRef;
}
