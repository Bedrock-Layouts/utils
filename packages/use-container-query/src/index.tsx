import { useForwardedRef } from "@bedrock-layout/use-forwarded-ref";
import { useResizeObserver } from "@bedrock-layout/use-resize-observer";
import React from "react";

interface UseContainerQueryProps {
  width?: number;
  maxWidth?: number;
}

export function useContainerQuery<T extends Element>(
  { width = 1, maxWidth }: Readonly<UseContainerQueryProps>,
  forwardedRef?: React.Ref<T>,
): [boolean, React.MutableRefObject<T>] {
  const [matches, setMatch] = React.useState(false);
  const containerRef = useForwardedRef(forwardedRef);

  useResizeObserver((entry: ResizeObserverEntry) => {
    //fix typings
    /* c8 ignore start */
    const nodeWidth =
      (entry.borderBoxSize as unknown as ResizeObserverSize)?.inlineSize ??
      entry.contentRect.width;
    /* c8 ignore end */
    //nodeWidth can be zero when it is switching from one node to another.  This will ignore that.
    if (nodeWidth > 0) {
      const newMatch =
        maxWidth === undefined
          ? nodeWidth <= width
          : nodeWidth >= width && nodeWidth <= maxWidth;

      setMatch(newMatch);
    }

    return entry;
  }, containerRef.current);

  return [matches, containerRef];
}
