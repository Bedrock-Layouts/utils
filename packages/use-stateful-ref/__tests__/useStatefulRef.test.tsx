import React from "react";
import ReactDOM from "react-dom/client";
import { act } from "react-dom/test-utils";
import { afterEach, beforeEach, describe, expect, it, test, vi } from "vitest";

import * as imports from "../src";

const { useStatefulRef } = imports;

vi.spyOn(React, "useState");

let state = undefined;

const setState = vi.fn((newState) => {
  state = newState;
});

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-expect-error
React.useState.mockImplementation((initialValue) => {
  state = initialValue;
  return [state, setState];
});

let statefulRef: React.MutableRefObject<unknown> | undefined;

const HookWrapper = ({
  value,
  initialValue,
}: {
  value: unknown;
  initialValue: unknown;
}) => {
  statefulRef = useStatefulRef(initialValue);
  if (value) {
    statefulRef.current = value;
  }
  return null;
};

describe("useStatefulRef", () => {
  let container: Node;

  beforeEach(() => {
    container = document.createElement("div");
    document.body.appendChild(container);
  });

  afterEach(() => {
    state = undefined;
    statefulRef = undefined;
    document.body.removeChild(container);
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-expect-error
    container = null;
    setState.mockRestore();
  });

  test("useStatefulRef is not null", () => {
    expect(useStatefulRef).toBeTruthy();
  });

  it("should call useState", () => {
    act(() => {
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-expect-error
      ReactDOM.createRoot(container).render(
        <HookWrapper value={undefined} initialValue={undefined} />,
      );
    });

    expect(React.useState).toBeCalled();
  });

  it("should call setState when setting value", () => {
    act(() => {
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-expect-error
      ReactDOM.createRoot(container).render(
        <HookWrapper initialValue="Testing" value="Test" />,
      );
    });

    expect(setState).toBeCalled();
    expect(statefulRef).toMatchSnapshot();
  });

  it("should call not call setState when setting value to same value", () => {
    expect(setState).not.toBeCalled();

    act(() => {
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-expect-error
      ReactDOM.createRoot(container).render(
        <HookWrapper initialValue="Test" value="Test" />,
      );
    });

    expect(setState).not.toBeCalled();

    expect(setState.mock.calls).toMatchSnapshot();
  });
});
