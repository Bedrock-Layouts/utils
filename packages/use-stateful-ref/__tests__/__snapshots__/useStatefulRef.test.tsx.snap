// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useStatefulRef > should call not call setState when setting value to same value 1`] = `[]`;

exports[`useStatefulRef > should call setState when setting value 1`] = `
{
  "current": "Test",
}
`;
